### PR TITLE
Make some amount of stack space available for API for threads using small stack sizes

### PR DESCRIPTION
**Pull request by kouvel**
- Previously, `TryEnsureSufficientExecutionStack` would always return false for a thread with slightly higher than 128 KB of stack space on 64-bit platforms
- Fixed to use a portion of the total stack space as a buffer with the same default as before for larger stack sizes

### CODE REVIEW

File: src/coreclr/vm/threads.cpp [hunk 1]
```diff
@@ -6431,18 +6431,21 @@ BOOL Thread::SetStackLimits(SetStackLimitScope scope)
...
-        const UINT_PTR MinExecutionStackSize = 128 * 1024;
+        const UINT_PTR DefaultMinExecutionStackSize = 128 * 1024;
 #else // !HOST_64BIT
-        const UINT_PTR MinExecutionStackSize = 64 * 1024;
+        const UINT_PTR DefaultMinExecutionStackSize = 64 * 1024;
 #endif // HOST_64BIT
-        _ASSERTE(m_CacheStackBase >= m_CacheStackLimit);
-        if ((reinterpret_cast<UINT_PTR>(m_CacheStackBase) - reinterpret_cast<UINT_PTR>(m_CacheStackLimit)) >
-            MinExecutionStackSize)
+        const UINT_PTR MinExecutionStackSize = Min(DefaultMinExecutionStackSize, Max<UINT_PTR>(MaxStackSize / 4, 1 * 1024));
```
Comment (jkotas) [inline src/coreclr/vm/threads.cpp:6447]: It does not sound right to me that TryEnsureSufficientExecutionStack reserve size would depend on the thread stack size. 

How are people supposed to test that their code will fit into the reserve when the behavior is changing based on the thread stack size?

Comment (kouvel) [reply src/coreclr/vm/threads.cpp:6447]: Currently, around 128 KB of thread stack size, the behavior changes significantly and suddenly. When a thread has slightly more than 128 KB of stack space, only that little portion will be considered and the API may always return false indicating that there is no space. With slightly below 128 KB of stack space, the API doesn't leave any buffer, always returning true and if more stack is used the app would crash.

Perhaps there could be a config var that specifies what the required min execution stack should be. Alternatively with the previous behavior, threads would need a bit more than 128 KB, say 256 KB to leave some room, to guarantee the default 128 KB of execution stack size on 64-bit. Thoughts?

Comment (jkotas) [reply src/coreclr/vm/threads.cpp:6447]: Is there are a issue or email describing the problem that this is trying to fix?

We can mention in the hosting documentation that the runtime performance can be degraded with stacks that are less than 256kB.

Comment (kouvel) [reply src/coreclr/vm/threads.cpp:6447]: There's a separate issue in `System.Linq.Expressions` that hits an effective infinite loop if `TryEnsureSufficientExecutionStack` always returns false. Filed https://github.com/dotnet/runtime/issues/61757, I'll go ahead and fix that instead and look into documenting this.

**Pull request closed without merging**